src/core/lib/event_engine/cf_engine/dns_service_resolver.cc

```diff
@@ -0,0 +1,205 @@
+// limitations under the License.
+
+#include <grpc/support/port_platform.h>
+
+#ifdef GPR_APPLE
+
+#include "absl/strings/str_format.h"
+
+#include "src/core/lib/address_utils/parse_address.h"
+#include "src/core/lib/event_engine/cf_engine/dns_service_resolver.h"
+#include "src/core/lib/event_engine/posix_engine/lockfree_event.h"
+#include "src/core/lib/event_engine/tcp_socket_utils.h"
+#include "src/core/lib/event_engine/trace.h"
+#include "src/core/lib/gprpp/host_port.h"
+
+namespace grpc_event_engine {
+namespace experimental {
+
+void DNSServiceResolverImpl::LookupHostname(
+    EventEngine::DNSResolver::LookupHostnameCallback on_resolve,
+    absl::string_view name, absl::string_view default_port) {
+  GRPC_EVENT_ENGINE_DNS_TRACE(
+      "DNSServiceResolverImpl::LookupHostname: name: %.*s, default_port: %.*s, "
+      "this: %p",
+      static_cast<int>(name.length()), name.data(),
+      static_cast<int>(default_port.length()), default_port.data(), this);
+
+  absl::string_view host;
+  absl::string_view port_string;
+  if (!grpc_core::SplitHostPort(name, &host, &port_string)) {
+    on_resolve(
+        absl::InvalidArgumentError(absl::StrCat("Unparseable name: ", name)));
+    return;
+  }
+  GPR_ASSERT(!host.empty());
+  if (port_string.empty()) {
+    if (default_port.empty()) {
+      on_resolve(absl::InvalidArgumentError(absl::StrFormat(
+          "No port in name %s or default_port argument", name)));
+      return;
+    }
+    port_string = default_port;
+  }
+
+  int port = 0;
+  if (port_string == "http") {
+    port = 80;
+  } else if (port_string == "https") {
+    port = 443;
+  } else if (!absl::SimpleAtoi(port_string, &port)) {
+    on_resolve(absl::InvalidArgumentError(
+        absl::StrCat("Failed to parse port in name: ", name)));
+    return;
+  }
+
+  // TODO(yijiem): Change this when refactoring code in
+  // src/core/lib/address_utils to use EventEngine::ResolvedAddress.
+  grpc_resolved_address addr;
+  const std::string hostport = grpc_core::JoinHostPort(host, port);
+  if (grpc_parse_ipv4_hostport(hostport.c_str(), &addr,
+                               true /* log errors */) ||
+      grpc_parse_ipv6_hostport(hostport.c_str(), &addr,
+                               true /* log errors */)) {
+    // Early out if the target is an ipv4 or ipv6 literal.
+    std::vector<EventEngine::ResolvedAddress> result;
+    result.emplace_back(reinterpret_cast<sockaddr*>(addr.addr), addr.len);
+    on_resolve(std::move(result));
+    return;
+  }
+
+  DNSServiceRef sdRef;
+  auto host_string = std::string{host};
+  auto error = DNSServiceGetAddrInfo(
+      &sdRef, kDNSServiceFlagsTimeout | kDNSServiceFlagsReturnIntermediates, 0,
+      kDNSServiceProtocol_IPv4 | kDNSServiceProtocol_IPv6, host_string.c_str(),
+      &DNSServiceResolverImpl::ResolveCallback, this /* do not Ref */);
+
+  if (error != kDNSServiceErr_NoError) {
+    on_resolve(absl::UnknownError(
```

Comment:
Calling this inline could be problematic. See my other comments on that.

I won't comment on them further, please locate and fix all instances.